### PR TITLE
Workaround for modded attributes causing warning on vanilla clients

### LINKED ISSUE #7368: [1.16] Forge Attributes cause error spamm on client
**Minecraft Version:** 1.16.3

**Forge Version:** 34.1.0

**Logs:** Vanilla client log: [https://pastebin.com/XVQfmuyQ](url)

**Steps to Reproduce:**
 1. Setup a forge server
 2. Connect to it with a vanilla client

**Description of issue:**
Since forge adds their own attributes (see ForgeMod.SWIM_SPEED, or ForgeMod.REACH_DISTANCE) when those get sent to the client for update the client obviously doesnt know about them causing "Entity x does not have attribute null" error message spamm

### PR DESCRIPTION
**Pull request by diesieben07**
When modded attributes are added to vanilla entities those cause warning logs on vanilla clients every time they receive the entity attributes.
This is a possible workaround.

Fixes #7368.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 1]
```diff
@@ -1210,4 +1213,20 @@ public static synchronized void updateBurns()
         modpacks.add("vanilla");
         return modpacks;
     }
+
+    public static IPacket<?> cleanPacketForPossibleVanillaConnection(ServerPlayNetHandler netHandler, IPacket<?> packet)
+    {
+        if (NetworkHooks.getConnectionType(() -> netHandler.netManager) == ConnectionType.VANILLA)
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Would this be cleaner using [`NetworkHooks.isVanillaConnection(netHandler.netManager)`](https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/main/java/net/minecraftforge/fml/network/NetworkHooks.java#L122)

File: src/main/resources/forge.sas [hunk 2]
```diff
@@ -106,4 +106,5 @@ net/minecraft/world/biome/BiomeAmbience func_235220_d_()Ljava/util/Optional; # g
 net/minecraft/world/biome/BiomeAmbience func_235222_e_()Ljava/util/Optional; # getAmbientSound
 net/minecraft/world/biome/BiomeAmbience func_235224_f_()Ljava/util/Optional; # getMoodSound
 net/minecraft/world/biome/BiomeAmbience func_235226_g_()Ljava/util/Optional; # getAdditionsSound
-net/minecraft/world/biome/BiomeAmbience func_235228_h_()Ljava/util/Optional; # getMusic
\ No newline at end of file
+net/minecraft/world/biome/BiomeAmbience func_235228_h_()Ljava/util/Optional; # getMusic
+net/minecraft/network/play/server/SEntityPropertiesPacket func_149441_d()Ljava/util/List; # getSnapshots
```
Comment (pupnewfster) [inline src/main/resources/forge.sas]: I believe the checkSAS task was forgotten here

File: src/main/java/net/minecraftforge/common/ForgeHooks.java (hunk 1 above)
Comment (diesieben07) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: You would be correct. Will change.

File: src/main/resources/forge.sas (hunk 2 above)
Comment (diesieben07) [reply src/main/resources/forge.sas]: Nope. I ran it. Ran it again just now, it did not change anything and did not report any errors.

Comment (pupnewfster) [reply src/main/resources/forge.sas]: Odd I thought the SAS file was supposed to be alphabetized just like the AT file.

Comment (diesieben07) [reply src/main/resources/forge.sas]: So did I.

Comment (LexManos) [reply src/main/resources/forge.sas]: It isn't done automatically. Its one of the todo's. But if you would be so kind, putting them in the correct place just makes it easer to read.

Comment (diesieben07) [reply src/main/resources/forge.sas]: It's ordered now, I didn't mean to come across as sassy

**Comment by diesieben07** [comment]
Lex: I have updated the patch like you suggested on Discord. Unfortunately the patch to the packet is still kinda large, but I am not sure if it could be done any better.

**Comment by cpw** [comment]
Suggesting @diesieben07 rewrite this using a netty pipeline handler interposed into the regular stream. This will allow for significantly more flexibility in handling vanilla connections, for a variety of scenarios, not just this one.

**Comment by diesieben07** [comment]
Switched to a netty pipeline handler approach as suggested. This significantly reduced the patch size (yay).

File: patches/minecraft/net/minecraft/network/handshake/ServerHandshakeNetHandler.java.patch [hunk 3]
```diff
@@ -1,10 +1,11 @@
 --- a/net/minecraft/network/handshake/ServerHandshakeNetHandler.java
 +++ b/net/minecraft/network/handshake/ServerHandshakeNetHandler.java
-@@ -23,6 +23,7 @@
+@@ -23,6 +23,8 @@
     }
  
     public void func_147383_a(CHandshakePacket p_147383_1_) {
 +      if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerLogin(p_147383_1_, this.field_147386_b)) return;
++      net.minecraftforge.network.VanillaConnectionNetworkFilter.injectIfNecessary(this.field_147386_b);
```
Comment (cpw) [inline patches/minecraft/net/minecraft/network/handshake/ServerHandshakeNetHandler.java.patch]: Can't this line be added to the method above? Why add it here and create a patch?

File: src/main/java/net/minecraftforge/network/VanillaConnectionNetworkFilter.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            manager.channel().pipeline().addBefore("packet_handler", "forge:vanilla_filter", new VanillaConnectionNetworkFilter());
+            LOGGER.debug("Injected into {}", manager);
+        }
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, IPacket<?> msg, List<Object> out)
+    {
+        IPacket<?> filtered;
+        if (msg instanceof SEntityPropertiesPacket)
```
Comment (cpw) [inline src/main/java/net/minecraftforge/network/VanillaConnectionNetworkFilter.java]: Can you make this a bit more pluggable? Put "handlers" as Functions into a map, and process the filter through that map. That will allow us to extend this behaviour to other packets we need to filter for -> vanillaclient traffic.

File: patches/minecraft/net/minecraft/network/handshake/ServerHandshakeNetHandler.java.patch (hunk 3 above)
Comment (diesieben07) [reply patches/minecraft/net/minecraft/network/handshake/ServerHandshakeNetHandler.java.patch]: I was under the impression you want to keep FML somewhat out of the "major game patches", so it seemed to me that FML calling into Forge code would be a bit strange. But I'll change it if that's no problem.

**Comment by diesieben07** [comment]
Switched to a handler based approach and simplified the patches.

**Comment by cpw** [comment]
This is a thing of beauty :+1:

**Review by cpw — APPROVED** [review]

**Comment by cpw** [comment]
acd18455c28d888afa95602b9ff1f6d1cca7ef6e

**Pull request closed without merging**